### PR TITLE
Raise argument error if template name matches multiple files

### PR DESCRIPTION
**Pull request by jcredding**
This updates the template source and null template engine logic
to raise an argument error if the template name matches multiple
template files. This also changes how the template source glob
for template files when parsing the ext to determine the engine it
should use. It no longer expects templates to have extensions and
no longer rejects if the templates extension is not a configured
extension. All of this is to try and make Sanfords template
rendering more straightforward and avoid surprises.

By raising an error if multiple template files match, this avoids
any surprises with an unexpected file being rendered. This is
particularly nefarious if a development machines orders files one
way when globbing and a production machine orders them a different
way. In development everything would work as expected but then
fail in production.

By not expecting a `.` after the template name and not rejecting
template files that do not match a configured engine, the
rendering is now more flexible. If a user wants to render a static
file (that doesn't need an engine) they can and the null template
engine will automatically take over for them. The template file
name also doesn't have to match an imposed convention. Finally,
this also sets up Sanford to allow using multiple template engines
in the future if needed.

Finally, this no longer disallows extensions for engine
registration. Previously, we would explicitly not allow rb engines
which would ensure that we didn't pick an rb file as a template.
This is really restrictive and doesn't really provide any
protection. If a user wants to use an template file with an rb
extension they should be allowed to.

@kellyredding - Ready for review.

### CODE REVIEW

File: lib/sanford/template_engine.rb [hunk 1]
```diff
@@ -30,12 +30,11 @@ def ==(other_engine)
...
-    def render(name, service_handler, locals)
-      template_file = self.source_path.join(name).to_s
-      unless File.exists?(template_file)
-        raise ArgumentError, "template file `#{template_file}` does not exist"
+    def render(template_name, service_handler, locals)
+      if (path = Dir.glob(self.source_path.join("#{template_name}*")).first).nil?
+        raise ArgumentError, "a template named `#{template_name}` does not exist"
       end
-      File.read(template_file)
+      File.read(path)
```
Comment (kellyredding) [inline lib/sanford/template_engine.rb]: @jcredding need to raise arg error if more than one globbed path here too.

File: test/unit/template_source_tests.rb [hunk 2]
```diff
@@ -127,6 +110,14 @@ class RenderTests < InitTests
       end
     end
 
+    should "raise an argument error if name matches multiple templates" do
+      # there should be 2 files called "template" in `test/support` with diff
```
Comment (kellyredding) [inline test/unit/template_source_tests.rb]: @jcredding I think you need to say "conflict_template" in this comment now.

**Comment by kellyredding** [comment]
@jcredding a couple comments - I'll look again.

File: lib/sanford/template_source.rb [hunk 3]
```diff
@@ -55,8 +47,11 @@ def ==(other_template_source)
     private
 
     def get_template_ext(template_name)
-      files = Dir.glob("#{File.join(@path, template_name.to_s)}.*")
-      files = files.reject{ |p| !@engines.keys.include?(parse_ext(p)) }
+      files = Dir.glob("#{File.join(@path, template_name.to_s)}*")
+      if files.size > 1
+        raise ArgumentError, "#{template_name.inspect} matches more than one " \
+                             "file, please use a more specific template name"
```
Comment (kellyredding) [inline lib/sanford/template_source.rb]: @jcredding maybe say "consider using a more...".

File: lib/sanford/template_source.rb [hunk 4]
```diff
@@ -55,8 +47,11 @@ def ==(other_template_source)
...
+      end
       parse_ext(files.first.to_s || '')
```
Comment (kellyredding) [inline lib/sanford/template_source.rb]: @jcredding don't think the `||` is needed.

File: lib/sanford/template_engine.rb [hunk 5]
```diff
@@ -30,12 +30,11 @@ def ==(other_engine)
 
   class NullTemplateEngine < TemplateEngine
 
-    def render(name, service_handler, locals)
-      template_file = self.source_path.join(name).to_s
-      unless File.exists?(template_file)
-        raise ArgumentError, "template file `#{template_file}` does not exist"
+    def render(template_name, service_handler, locals)
+      if (path = Dir.glob(self.source_path.join("#{template_name}*")).first).nil?
+        raise ArgumentError, "a template named `#{template_name}` does not exist"
```
Comment (kellyredding) [inline lib/sanford/template_engine.rb]: @jcredding we discussed using `.inspect` over back-ticks here.

File: lib/sanford/template_source.rb (hunk 4 above)
Comment (kellyredding) [inline lib/sanford/template_source.rb]: @jcredding also `.to_s` probably isn't needed since coming from a glob.

Comment (kellyredding) [inline lib/sanford/template_source.rb]: @jcredding nevermind, globs can return `nil`

**Comment by kellyredding** [comment]
@jcredding this all looks good - thanks! :-1:

File: test/unit/template_engine_tests.rb [hunk 6]
```diff
@@ -70,11 +70,18 @@ class NullTemplateEngineTests < Assert::Context
...
 
     should "read and return the given path in its source path on `render" do
-      exists_file = 'test/support/template.json'
+      exists_file = 'test/support/template.erb'
       exp = File.read(subject.source_path.join(exists_file).to_s)
       assert_equal exp, subject.render(exists_file, @service_handler, @locals)
     end
 
+    should "complain if given a path that matches multiple files" do
+      no_exists_file = 'test/support/conflict_template'
```
Comment (kellyredding) [inline test/unit/template_engine_tests.rb]: @jcredding maybe update this to do a choice on full file name or partial file name.  Then use glob* to get the path to read from.  We should do this to test that it works on both full and partial file names.

File: test/unit/template_engine_tests.rb [hunk 7]
```diff
@@ -70,11 +70,18 @@ class NullTemplateEngineTests < Assert::Context
...
     should "read and return the given path in its source path on `render" do
-      exists_file = 'test/support/template.json'
-      exp = File.read(subject.source_path.join(exists_file).to_s)
+      exists_file = ['test/support/template', 'test/support/template.erb'].choice
+      exp = File.read(Dir.glob("#{subject.source_path.join(exists_file)}*").first)
       assert_equal exp, subject.render(exists_file, @service_handler, @locals)
     end
 
+    should "complain if given a path that matches multiple files" do
+      no_exists_file = 'test/support/conflict_template'
```
Comment (kellyredding) [inline test/unit/template_engine_tests.rb]: @jcredding maybe call this conflict file.

File: test/unit/template_source_tests.rb [hunk 8]
```diff
@@ -127,6 +110,14 @@ class RenderTests < InitTests
       end
     end
 
+    should "raise an argument error if name matches multiple templates" do
```
Comment (kellyredding) [inline test/unit/template_source_tests.rb]: @jcredding maybe say "complain if the given template name matches..." - better matches other "complain" test descriptions.

File: lib/sanford/template_engine.rb [hunk 9]
```diff
@@ -30,12 +30,16 @@ def ==(other_engine)
...
-      unless File.exists?(template_file)
-        raise ArgumentError, "template file `#{template_file}` does not exist"
+    def render(template_name, service_handler, locals)
+      paths = Dir.glob(self.source_path.join("#{template_name}*"))
+      if paths.size > 1
+        raise ArgumentError, "#{template_name.inspect} matches more than one " \
+                             "file, consider using a more specific template name"
+      end
+      if paths.size < 1
+        raise ArgumentError, "a template named #{template_name.inspect} does not exist"
```
Comment (kellyredding) [inline lib/sanford/template_engine.rb]: @jcredding I'm using this error message in Deas, FYI: "a template file named with ...".  Just trying to be specific about how the template lookups work.

See also:
* https://github.com/redding/deas-erubis/pull/27/files#diff-76e3cfcec93d9c7ca3a112a5bf6c9a45R51
* https://github.com/redding/nm/pull/17/files#diff-461c530e59120e9d700d5735752898e2R33

**Comment by kellyredding** [comment]
@jcredding sorry for all the rando comments on this.  I was re-reading the commit msg and something didn't seem clear.  What did you mean by "It no longer expects the template name to proceed an
extension ..."?  Did you mean "*precede*".  If so, maybe a clearer way to say it is "It no longer expects templates to have extensions ..."?

**Comment by jcredding** [comment]
@kellyredding - I've updated this with your latest comments.

**Comment by kellyredding** [comment]
@jcredding this looks good to me now - thanks! :-1:

**Pull request merged**